mobly/controllers/android_device.py

```diff
@@ -207,15 +217,16 @@ def get_instances_with_configs(configs):
             raise Error(
                 "Required value 'serial' is missing in AndroidDevice config %s."
                 % c)
-        is_required = c.pop("required", True)
+        is_required = c.get(KEY_DEVICE_REQUIRED, True)
         try:
             ad = AndroidDevice(serial)
             ad.load_config(c)
-        except:
-            if not is_required:
-                logging.warning("Failed to initialize Android device %s", serial)
-                continue
-            raise
+        except Exception as e:
+            if is_required:
+                raise
+            logging.warning("Failed to initialize optional device %s: %s",
+                            serial, e)
+            continue
```

Comment:
I think we wanted to append None instead of continuing, right? Otherwise you can't find out how many devices you're supposed to have and which ones failed to come up. It also makes the indexes wrong:
"AndroidDevice": ["A", "B", "C"]

so you'd expect ad[2] to be C, but if B failed to come up it's actually [1].